### PR TITLE
Clarify vec docs on deallocation (fixes #46879)

### LINKED ISSUE #46879: Clarify confusing wording on Vec
Originally reported here: https://news.ycombinator.com/item?id=15969317

----

I don't understand the following sentence:

> ... it is strongly recommended that you only free memory allocated by a Vec by creating a new Vec and dropping it.

Could someone parse this for me? To me, it reads like "to free memory allocated by a Vec v, create a new Vec w and drop it (i.e., w)".

This is obviously not what it is intended. What's the role of the second Vec? Is this only in a context where v is empty but its memory not yet freed? If so, why wouldn't shrink_to_fit work in that case?

### PR DESCRIPTION
**Pull request by Manishearth**
r? @steveklabnik

### CODE REVIEW

File: src/liballoc/vec.rs [hunk 1]
```diff
@@ -224,8 +224,10 @@ use Bound::{Excluded, Included, Unbounded};
 /// types inside a `Vec`, it will not allocate space for them. *Note that in this case
 /// the `Vec` may not report a [`capacity`] of 0*. `Vec` will allocate if and only
 /// if [`mem::size_of::<T>`]`() * capacity() > 0`. In general, `Vec`'s allocation
-/// details are subtle enough that it is strongly recommended that you only
-/// free memory allocated by a `Vec` by creating a new `Vec` and dropping it.
+/// details are very subtle &mdash; if you intend to allocate memory using a `Vec`
+/// and use it for something else (either to pass to unsafe code, or to build your
+/// own memory-backed collection), be sure to deallocate this memory by using
+/// `from_raw_parts` to recover the `Vec` and then dropping it.
```
Comment (kjeremy) [inline src/liballoc/vec.rs]: As a newb this is confusing.  Maybe an explicit inline example would be good.

**Comment by tinaun** [comment]
“to recover the Vec and then dropping it” sounds weird but I don’t know if “drop it” is grammatically correct either

**Review by steveklabnik — APPROVED** [review]

**Comment by steveklabnik** [comment]
@bors: r+ rollup

@tinaun it doesn't sound odd to me; we might be able to simplify it by splitting the sentences up a bit, but this is good enough to fix the bug, imo.

**Comment by Manishearth** [comment]
"drop it" is grammatically correct; besides, "drop" is a technical term in
rust so we us it a bit differently :)

**Pull request merged**